### PR TITLE
Added changes for password validations and error messages.

### PR DESCRIPTION
**Pull request by pravalikarouthu**
# Password and Email Validations

#### Please check if the PR fulfills these requirements

- [x] The changes don't come from your master branch. ([Source](https://blog.jasonmeridth.com/posts/do-not-issue-pull-requests-from-your-master-branch/)) 

---

#### What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)
Introducing validation changes for the email, password(donor/client registration) and we are returning 400 status code when validations fail. Also, return error messages for failed validations while updating the user(donor/client) details.

#### What is the current behavior? (You can also link to an open issue here)
The status code that was been returned was 422, which was wrong. Also, validation errors were not sent to the UI on updates.

#### What is the new behavior? (if this is a feature change)
We are returning 400 status code when validations fail. Also, return error messages for failed validations while updating the user(donor/client) details. 

#### Does this PR introduce a breaking change? (What changes might users need to make in their application due to this PR?)
No

#### Images (before/ after screenshots, interaction GIFs, ...)

---

#### TODO
 - this
 - that
 - the other

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -49,3 +49,6 @@ gem "faker", "~> 2.7"
 gem "geocoder", "~> 1.5.2"
 
 gem "httparty", "~> 0.17.3"
+
+
+
```
Comment (sagehen03) [inline Gemfile:54]: Looks like you have some unrelated changes in Gemfile and Gemfile.lock.

File: app/controllers/client_auth_controller.rb [hunk 2]
```diff
@@ -1,5 +1,6 @@
 class ClientAuthController < ApplicationController
 	skip_before_action :authorized, only: [:create]
+
```
Comment (sagehen03) [inline app/controllers/client_auth_controller.rb:3]: Not a big deal at all but it looks like you've added some extra whitespace here.  I'd be in favor of keeping the new line at the end but not the extra line at the top.

File: app/controllers/clients_controller.rb [hunk 3]
```diff
@@ -1,5 +1,5 @@
 class ClientsController < ApplicationController
-  skip_before_action :authorized, only: [:create, :update]
+    skip_before_action :authorized, only: [:create]
```
Comment (sagehen03) [inline app/controllers/clients_controller.rb:2]: It looks like some extra indentation snuck in here.  I don't think it matters in terms of the code working, but it would be nice to keep consistent indentation.

Comment (pravalikarouthu) [reply app/controllers/clients_controller.rb:2]: > Hey Pravalika, I think this is a good start, but I have some requests for you. Can you add some unit tests to verify that all our validation are working as expected? I've created a branch called `password-validation-changes-pr` that has an example of a test. Also can we use `client|donor.errors.messages` rather than mapping the rails error object the `failure_message`? There are already some tests for the donor and client controllers. You should verify that those are working too. If this seems like too much, I can go ahead and make the changes for one of the donor or client and you can use that as the template for the other. Happy to talk on slack about this too. Thanks!

Hey Drew!! That's really a good point and I totally agree about adding unit tests. I can go ahead and work on the unit tests part. I'll update you regarding the error messages by tomorrow!!

**Comment by pravalikarouthu** [comment]
> Hey Pravalika, I think this is a good start, but I have some requests for you. Can you add some unit tests to verify that all our validation are working as expected? I've created a branch called `password-validation-changes-pr` that has an example of a test. Also can we use `client|donor.errors.messages` rather than mapping the rails error object the `failure_message`? There are already some tests for the donor and client controllers. You should verify that those are working too. If this seems like too much, I can go ahead and make the changes for one of the donor or client and you can use that as the template for the other. Happy to talk on slack about this too. Thanks!

Hey Drew!! That's a good point. Definitely it's a good idea to add some unit tests. I'll go ahead and add the unit tests and also see what changes I can make regarding the 'failure_message' code. Thank you!!

**Comment by pravalikarouthu** [comment]
> Hey Pravalika, I think this is a good start, but I have some requests for you. Can you add some unit tests to verify that all our validation are working as expected? I've created a branch called `password-validation-changes-pr` that has an example of a test. Also can we use `client|donor.errors.messages` rather than mapping the rails error object the `failure_message`? There are already some tests for the donor and client controllers. You should verify that those are working too. If this seems like too much, I can go ahead and make the changes for one of the donor or client and you can use that as the template for the other. Happy to talk on slack about this too. Thanks!

Hi Drew!

Regarding the changes for the error messages, I think the error messages that I have are more meaningful than the default error messages. Here is a screenshot containing the comparisons between the two:

Screen Shot 2020-06-25 at 1 21 31 PM
Screen Shot 2020-06-25 at 1 22 10 PM

If you notice the standard rails messages, they don't tell the client what fields are inaccurate whereas the latter does that. I also changed the update API only because create API is not authenticated at the moment and can be called by anyone without authorization. We may have to talk about this in another conversation.
